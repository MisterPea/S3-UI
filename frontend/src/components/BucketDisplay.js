import * as React from 'react';
import { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { motion, AnimatePresence } from 'framer-motion';
import { IoAddCircleSharp } from 'react-icons/io5';
import LoadingBar from './graphic_elements/LoadingBar';
import { getBucketList } from '../redux/actions/bucket';
import ModalComponentWrapper from './ModalComponentWrapper';
import AddBucketModal from './AddBucketModal';
import BucketUL from './BucketUL';
import useScrollIntersect from './helpers/useScrollIntersect';

export default function BucketDisplay() {
  const { loading, buckets } = useSelector((state) => state);
  const [addBucketModal, setAddBucketModal] = useState(false);
  const [[scrollTarget], setScrollTarget] = useState([null]);
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(getBucketList());
  }, []);

  const [
    topScrollShadow,
    bottomScrollShadow,
    handleScroll] = useScrollIntersect(buckets, scrollTarget);

  function handleToggleModal() {
    setAddBucketModal((s) => !s);
  }

  function handleKeyPress(e) {
    e.preventDefault();
    if (e.key === 'Enter') {
      handleToggleModal();
    }
  }

  const loaderVariant = {
    exit: {
      opacity: 0,
      transition: {
        ease: 'circOut',
        when: 'beforeChildren',
        duration: 0.3,
      },
    },
  };

  const bucketVariant = {
    open: {
      opacity: 1,
      transition: {
        ease: 'easeIn',

      },
    },
    closed: {
      opacity: 0,
      transition: {
        // delay: 0.5,
      },
    },
  };

  return (
    <div className="bucket-display-wrapper">
      <div className="bucket-header-wrapper">
        <div className={`bucket-display ${topScrollShadow ? 'overflow' : ''}`}>
          <h3 className="bucket-name-header">Bucket Name</h3>
          <h3 className="bucket-date-header">Creation Date</h3>
          <h3 className="bucket-region-header">AWS Region</h3>
        </div>
      </div>
      <AnimatePresence exitBeforeEnter>
        {loading && (
          <motion.div key="loadingBar" variants={loaderVariant} exit="exit">
            <LoadingBar key="loadingBar" />
          </motion.div>
        )}
        <motion.div
          onScroll={(e) => handleScroll(e)}
          className="ul-wrapper"
          key="bucketUL"
          variants={bucketVariant}
          animate="open"
          initial="closed"
          exit="closed"
        >
          <BucketUL buckets={buckets} loading={!loading} onFinish={setScrollTarget} />
        </motion.div>
      </AnimatePresence>

      <div
        onClick={handleToggleModal}
        onKeyPress={(e) => handleKeyPress(e)}
        role="button"
        tabIndex={0}
        className={`add-bucket-bar${bottomScrollShadow ? ' overflow' : ''}`}
      >
        <span className="bucket-button-elements">
          <div className="bucket-cta-wrapper">
            <h3>Create New S3 Bucket</h3>
            <IoAddCircleSharp className="add-bucket-plus" />
          </div>
        </span>
      </div>

      {addBucketModal
        && (
        <ModalComponentWrapper close={handleToggleModal}>
          <AddBucketModal />
        </ModalComponentWrapper>
        )}
    </div>
  );
}
