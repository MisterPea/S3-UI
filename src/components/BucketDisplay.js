import * as React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { IoAddCircleSharp } from 'react-icons/io5';
import BucketLogo from './BucketLogo';
import { getBucketContentsList } from '../actions/bucket';

/**
 * Component to render bucket and info
 * @return {JSX}
 */
export default function BucketDisplay() {
  const buckets = useSelector((state) => state.buckets);
  const dispatch = useDispatch();
  function handleBucketClick(name, region) {
    dispatch(getBucketContentsList(region, name));
  }

  function onKeyDownHandler(e, name, region) {
    if (e.keyCode === 13) {
      handleBucketClick(name, region);
    }
  }

  return (
    <>
      <div className="table-head">
        <h3 className="name-header">Bucket Name</h3>
        <h3 className="date-header">Creation Date</h3>
        <h3 className="region-header">AWS Region</h3>
      </div>

      <ul>
        {buckets.map(({ Name, CreationDate, Region }) => (
          <li key={Name}>
            <div
              className="bucket-row"
              role="button"
              tabIndex={0}
              onClick={() => handleBucketClick(Name, Region)}
              onKeyDown={(e) => onKeyDownHandler(e, Name, Region)}
            >
              <div className="bucket-table-data name">
                <div className="bucket-logo"><BucketLogo /></div>
                <p className="bucket-name">{Name}</p>
              </div>
              <p className="bucket-table-data date">{CreationDate}</p>
              <p className="bucket-table-data region">{Region}</p>
            </div>
          </li>
        ))}
      </ul>

      <div className="add-bucket-bar">
        <span className="bucket-button-elements">
          <div className="bucket-cta-wrapper">
            <h3>Creat New S3 Bucket</h3>
            <IoAddCircleSharp className="add-bucket-plus" />
          </div>
        </span>
      </div>
    </>
  );
}
