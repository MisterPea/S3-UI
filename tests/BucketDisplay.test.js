import * as React from 'react';
import ReactDOM from 'react-dom';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import * as reactRedux from 'react-redux';
import BucketDisplay from '../src/components/BucketDisplay';

configure({ adapter: new Adapter() });

const buckets = [
  { Name: '1.newBucket', Region: 'us-east-1', CreationDate: '2021-06-15T15:50:10.000Z' },
  { Name: 'cf-templates-1ringo9ne14-us-east-1', Region: 'us-east-1', CreationDate: '2020-08-11T00:22:08.000Z' },
  { Name: 'elasticbeanstalk-us-east-1-897278278762', Region: 'us-east-1', CreationDate: '2020-08-13T18:36:57.000Z' },
  { Name: 'logs.misterpea.me', Region: 'us-east-2', CreationDate: '2020-02-23T21:01:23.000Z' },
  { Name: 'misterpea.me', Region: 'us-east-2', CreationDate: '2020-02-23T19:57:48.000Z' },
  { Name: 'misterpea.me-emails', Region: 'us-east-1', CreationDate: '2021-01-25T03:06:10.000Z' },
  { Name: 'new-ass-bucket', Region: 'us-east-1', CreationDate: '2021-04-15T13:48:06.000Z' },
  { Name: 'topStories.misterpea.me', Region: 'us-east-2', CreationDate: '2020-04-09T22:56:59.000Z' },
  { Name: 'www.misterpea.me', Region: 'us-east-2', CreationDate: '2020-02-23T19:59:01.000Z' },
];

describe('Bucket Display Component', () => {
  const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
  useSelectorMock.mockReturnValue(buckets);

  it('Renders without crashing', () => {
    const div = document.createElement('DIV');
    ReactDOM.render(<BucketDisplay />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('Renders the content of buckets array', () => {
    const wrapper = shallow(<BucketDisplay />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
