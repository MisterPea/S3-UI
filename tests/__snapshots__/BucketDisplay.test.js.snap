// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bucket Display Component Renders the content of buckets array 1`] = `
<Fragment>
  <div
    className="buckets-number"
  >
    Buckets (9)
  </div>
  <div
    className="table-head bucket-display"
  >
    <h3
      className="name-header"
    >
      Bucket Name
    </h3>
    <h3
      className="date-header"
    >
      Creation Date
    </h3>
    <h3
      className="region-header"
    >
      AWS Region
    </h3>
  </div>
  <ul>
    <li
      key="1.newBucket"
    >
      <Link
        to="/S3/?id=1.newBucket&loc=us-east-1"
      >
        <div
          className="bucket-row"
          role="button"
          tabIndex={0}
        >
          <div
            className="bucket-table-data name"
          >
            <div
              className="bucket-logo"
            >
              <BucketLogo />
            </div>
            <p
              className="bucket-name"
            >
              1.newBucket
            </p>
          </div>
          <p
            className="bucket-table-data date"
          >
            2021-06-15T15:50:10.000Z
          </p>
          <p
            className="bucket-table-data region"
          >
            us-east-1
          </p>
        </div>
      </Link>
    </li>
    <li
      key="cf-templates-1ringo9ne14-us-east-1"
    >
      <Link
        to="/S3/?id=cf-templates-1ringo9ne14-us-east-1&loc=us-east-1"
      >
        <div
          className="bucket-row"
          role="button"
          tabIndex={0}
        >
          <div
            className="bucket-table-data name"
          >
            <div
              className="bucket-logo"
            >
              <BucketLogo />
            </div>
            <p
              className="bucket-name"
            >
              cf-templates-1ringo9ne14-us-east-1
            </p>
          </div>
          <p
            className="bucket-table-data date"
          >
            2020-08-11T00:22:08.000Z
          </p>
          <p
            className="bucket-table-data region"
          >
            us-east-1
          </p>
        </div>
      </Link>
    </li>
    <li
      key="elasticbeanstalk-us-east-1-897278278762"
    >
      <Link
        to="/S3/?id=elasticbeanstalk-us-east-1-897278278762&loc=us-east-1"
      >
        <div
          className="bucket-row"
          role="button"
          tabIndex={0}
        >
          <div
            className="bucket-table-data name"
          >
            <div
              className="bucket-logo"
            >
              <BucketLogo />
            </div>
            <p
              className="bucket-name"
            >
              elasticbeanstalk-us-east-1-897278278762
            </p>
          </div>
          <p
            className="bucket-table-data date"
          >
            2020-08-13T18:36:57.000Z
          </p>
          <p
            className="bucket-table-data region"
          >
            us-east-1
          </p>
        </div>
      </Link>
    </li>
    <li
      key="logs.misterpea.me"
    >
      <Link
        to="/S3/?id=logs.misterpea.me&loc=us-east-2"
      >
        <div
          className="bucket-row"
          role="button"
          tabIndex={0}
        >
          <div
            className="bucket-table-data name"
          >
            <div
              className="bucket-logo"
            >
              <BucketLogo />
            </div>
            <p
              className="bucket-name"
            >
              logs.misterpea.me
            </p>
          </div>
          <p
            className="bucket-table-data date"
          >
            2020-02-23T21:01:23.000Z
          </p>
          <p
            className="bucket-table-data region"
          >
            us-east-2
          </p>
        </div>
      </Link>
    </li>
    <li
      key="misterpea.me"
    >
      <Link
        to="/S3/?id=misterpea.me&loc=us-east-2"
      >
        <div
          className="bucket-row"
          role="button"
          tabIndex={0}
        >
          <div
            className="bucket-table-data name"
          >
            <div
              className="bucket-logo"
            >
              <BucketLogo />
            </div>
            <p
              className="bucket-name"
            >
              misterpea.me
            </p>
          </div>
          <p
            className="bucket-table-data date"
          >
            2020-02-23T19:57:48.000Z
          </p>
          <p
            className="bucket-table-data region"
          >
            us-east-2
          </p>
        </div>
      </Link>
    </li>
    <li
      key="misterpea.me-emails"
    >
      <Link
        to="/S3/?id=misterpea.me-emails&loc=us-east-1"
      >
        <div
          className="bucket-row"
          role="button"
          tabIndex={0}
        >
          <div
            className="bucket-table-data name"
          >
            <div
              className="bucket-logo"
            >
              <BucketLogo />
            </div>
            <p
              className="bucket-name"
            >
              misterpea.me-emails
            </p>
          </div>
          <p
            className="bucket-table-data date"
          >
            2021-01-25T03:06:10.000Z
          </p>
          <p
            className="bucket-table-data region"
          >
            us-east-1
          </p>
        </div>
      </Link>
    </li>
    <li
      key="new-ass-bucket"
    >
      <Link
        to="/S3/?id=new-ass-bucket&loc=us-east-1"
      >
        <div
          className="bucket-row"
          role="button"
          tabIndex={0}
        >
          <div
            className="bucket-table-data name"
          >
            <div
              className="bucket-logo"
            >
              <BucketLogo />
            </div>
            <p
              className="bucket-name"
            >
              new-ass-bucket
            </p>
          </div>
          <p
            className="bucket-table-data date"
          >
            2021-04-15T13:48:06.000Z
          </p>
          <p
            className="bucket-table-data region"
          >
            us-east-1
          </p>
        </div>
      </Link>
    </li>
    <li
      key="topStories.misterpea.me"
    >
      <Link
        to="/S3/?id=topStories.misterpea.me&loc=us-east-2"
      >
        <div
          className="bucket-row"
          role="button"
          tabIndex={0}
        >
          <div
            className="bucket-table-data name"
          >
            <div
              className="bucket-logo"
            >
              <BucketLogo />
            </div>
            <p
              className="bucket-name"
            >
              topStories.misterpea.me
            </p>
          </div>
          <p
            className="bucket-table-data date"
          >
            2020-04-09T22:56:59.000Z
          </p>
          <p
            className="bucket-table-data region"
          >
            us-east-2
          </p>
        </div>
      </Link>
    </li>
    <li
      key="www.misterpea.me"
    >
      <Link
        to="/S3/?id=www.misterpea.me&loc=us-east-2"
      >
        <div
          className="bucket-row"
          role="button"
          tabIndex={0}
        >
          <div
            className="bucket-table-data name"
          >
            <div
              className="bucket-logo"
            >
              <BucketLogo />
            </div>
            <p
              className="bucket-name"
            >
              www.misterpea.me
            </p>
          </div>
          <p
            className="bucket-table-data date"
          >
            2020-02-23T19:59:01.000Z
          </p>
          <p
            className="bucket-table-data region"
          >
            us-east-2
          </p>
        </div>
      </Link>
    </li>
  </ul>
  <div
    className="add-bucket-bar"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <span
      className="bucket-button-elements"
    >
      <div
        className="bucket-cta-wrapper"
      >
        <h3>
          Create New S3 Bucket
        </h3>
        <IoAddCircleSharp
          className="add-bucket-plus"
        />
      </div>
    </span>
  </div>
</Fragment>
`;
